base/get.go

```diff
@@ -26,17 +31,81 @@ func (i GetHttpRequestInput) GetValidStatusFunc() ValidStatusFunc {
 // Get performs a GET request.
 func (c Client) Get(ctx context.Context, input GetHttpRequestInput) (*http.Response, int, error) {
 	var status int
-	url, err := c.buildUri(input.Uri)
-	if err != nil {
-		return nil, status, fmt.Errorf("unable to make request: %v", err)
+
+	// Check for a raw uri, else build one from the Uri field
+	url := input.rawUri
+	if url == "" {
+		var err error
+		url, err = c.buildUri(input.Uri)
+		if err != nil {
+			return nil, status, fmt.Errorf("unable to make request: %v", err)
+		}
 	}
+
+	// Build a new request
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, http.NoBody)
 	if err != nil {
 		return nil, status, err
 	}
+
+	// Perform the request
 	resp, status, err := c.performRequest(req, input)
 	if err != nil {
 		return nil, status, err
 	}
+
+	// Check for json content before handling pagination
+	contentType := strings.ToLower(resp.Header.Get("Content-Type"))
+	if strings.HasPrefix(contentType, "application/json") {
+		// Read the response body and close it
+		respBody, _ := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+
+		// Unmarshall odata
+		var odata OData
+		if err := json.Unmarshal(respBody, &odata); err != nil {
+			return nil, status, err
+		}
+
+		if odata.NextLink == nil || odata.Value == nil {
+			// No more pages, reassign response body and return
+			resp.Body = ioutil.NopCloser(bytes.NewBuffer(respBody))
+			return resp, status, nil
+		}
+
+		// Get the next page, recursively
+		nextInput := input
+		nextInput.rawUri = *odata.NextLink
+		nextResp, status, err := c.Get(ctx, nextInput)
```

Comment:
@manicminer my very basic understanding is that Golang doesn't have Tail Call Optimization which may (?) cause issues with recursive functions.

I don't know much about this, but been told by co-workers and heard it on a podcast.

Found this on SO: https://stackoverflow.com/questions/12102675/tail-call-optimization-in-go

Maybe a for loop should be used, or expose methods to the end users with something like Next() and let the end user page themselves?